soroban-env-common/src/raw_val.rs

```diff
@@ -1,101 +1,163 @@
 
-/// Code values for the 3 "tag" bits in the bit-packed representation
-/// of [RawVal].
+/// Code values for the 8 `tag` bits in the bit-packed representation
+/// of [RawVal]. These don't coincide with tag numbers in the SCVal XDR
+/// but cover all those cases as well as some optimized refinements for
+/// special cases (boolean true and false, small-value forms).
 #[repr(u8)]
-#[derive(Copy, Clone, PartialEq, Eq, PartialOrd, Ord, Hash)]
+#[derive(Copy, Clone, PartialEq, Eq, PartialOrd, Ord, Hash, Debug)]
 pub enum Tag {
+    /// Tag for a [RawVal] that encodes [bool] `false`. The bool type is refined to
+    /// two single-value subtypes in order for each tag number to coincides with
+    /// the WASM encoding of a boolean.
+    False = 0,
+
+    /// Tag for a [RawVal] that encodes [bool] `true`.
+    True = 1,
+
+    /// Tag for a [RawVal] that is empty/absent (eg. void, null, nil, undefined, None)
+    Void = 2,
+
+    /// Tag for a [RawVal] that is contains an error code.
+    Status = 3,
+
     /// Tag for a [RawVal] that contains a [u32] number.
-    U32 = 0,
+    U32 = 4,
 
     /// Tag for a [RawVal] that contains an [i32] number.
-    I32 = 1,
+    I32 = 5,
+
+    /// Tag for a [RawVal] that contains a [u64] small enough to fit in 56 bits.
+    U64Small = 6,
+
+    /// Tag for a [RawVal] that contains an [i64] small enough to fit in 56 bits.
+    I64Small = 7,
+
+    /// Tag for a [RawVal] that contains a [u64] timepoint small enough to fit
+    /// in 56 bits.
+    TimepointSmall = 8,
+
+    /// Tag for a [RawVal] that contains a [u64] duration small enough to fit in
+    /// 56 bits.
+    DurationSmall = 9,
+
+    /// Tag for a [RawVal] that contains a [u128] small enough to fit in 56 bits.
+    U128Small = 10,
+
+    /// Tag for a [RawVal] that contains a [i128] small enough to fit in 56 bits.
+    I128Small = 11,
+
+    /// Tag for a [RawVal] that contains a [u256] small enough to fit in 56 bits.
+    U256Small = 12,
+
+    /// Tag for a [RawVal] that contains a [i256] small enough to fit in 56 bits.
+    I256Small = 13,
+
+    /// Tag for a [RawVal] that contains up to 9 character symbols.
+    SymbolSmall = 14,
+
+    /// Tag for a [RawVal] that corresponds to
+    /// [stellar_xdr::ScVal::LedgerKeyContractExecutable]
+    LedgerKeyContractExecutable = 15,
 
-    /// Tag for a [RawVal] that contains a "static" value like `true`, `false`, `void`; see [Static](crate::Static).
-    Static = 2,
+    /// Code delimiting the upper boundary of "small" types.
+    SmallCodeUpperBound = 16,
 
-    /// Tag for a [RawVal] that contains a host object handle; see [Object](crate::Object).
-    Object = 3,
+    /// Tag reserved to indicate boundary between tags for "small" types with
+    /// their payload packed into the remaining 56 bits of the [RawVal] and
+    /// "object" types that are stored as host objects and referenced by
+    /// [Object] handle.
+    ObjectCodeLowerBound = 63,
```

Comment:
Are `ObjectCodeLowerBound` and `ObjectCodeUpperBound` just static values added here for convenience? Or do they actually represent valid `Val` cases?

Reply:
A little more than convenience -- we compare against them and need to be careful not to cast to values not-defined as enum cases -- but no they do not correspond to ScVal cases.